addons/binding/org.openhab.binding.nuki/src/main/java/org/openhab/binding/nuki/internal/NukiHandlerFactory.java

```diff
@@ -0,0 +1,140 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.nuki.internal;
+
+import java.util.Hashtable;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.servlet.http.HttpServlet;
+
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.smarthome.core.net.HttpServiceUtil;
+import org.eclipse.smarthome.core.net.NetworkAddressService;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory;
+import org.eclipse.smarthome.io.net.http.HttpClientFactory;
+import org.openhab.binding.nuki.NukiBindingConstants;
+import org.openhab.binding.nuki.handler.NukiBridgeHandler;
+import org.openhab.binding.nuki.handler.NukiSmartLockHandler;
+import org.openhab.binding.nuki.internal.dataexchange.NukiApiServlet;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.http.HttpService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link NukiHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Markus Katter - Initial contribution
+ */
+@Component(service = { ThingHandlerFactory.class, NukiHandlerFactory.class }, configurationPid = "binding.nuki")
+public class NukiHandlerFactory extends BaseThingHandlerFactory {
+
+    private final Logger logger = LoggerFactory.getLogger(NukiHandlerFactory.class);
+
+    private HttpService httpService;
+    private HttpClient httpClient;
+    private NetworkAddressService networkAddressService;
+    private Map<ThingUID, ServiceRegistration<NukiApiServlet>> nukiApiServletRegs = new ConcurrentHashMap<>();
+    private String openHabIpAndPort;
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return NukiBindingConstants.SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    @Override
+    protected ThingHandler createHandler(Thing thing) {
+        logger.debug("NukiHandlerFactory:createHandler({})", thing);
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (NukiBindingConstants.THING_TYPE_BRIDGE_UIDS.contains(thingTypeUID)) {
+            NukiApiServlet nukiApiServlet = new NukiApiServlet(httpService);
```

Comment:
Ok, this all looks pretty convoluted, I think we have to do some cleanup wrt the servlet architecture.
- First of all, there is no need to register the servlets as an OSGi service - it even is counter-productive as some OSGi frameworks might use the whiteboard pattern for servlets and you end up having it registered twice.
- Currently, the handler gets the servlet injected and the servlet gets hold of the handler through the Thing. So there's a circular dependency, which does not seem right. Imho, only the servlet requires the handler - there is no need to pass the servlet to the handler as it isn't doing anything with it.
- Servlets are rather "expensive", so imho it would be nicer to register a single servlet here in the factory that serves as the only implementation to deal with incoming callbacks.
- The factory could simply register newly created handlers at this single servlet instance together with their id that is used for the callback. The servlet can then easily dispatch incoming requests to the appropriate handler instance (or log an error if there's a request for a UID that does not exist).

Does that make sense to you?